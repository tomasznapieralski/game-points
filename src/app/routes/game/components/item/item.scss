.item {
  display: flex;
  justify-content: center;
  align-items: center;
  font-size: inherit;

  &__icon {
    &--biohazard {
      color: #af00af;
    }

    &--fire {
      color: #e20000;
    }

    &--spider {
      color: #000000;
    }

    &--lemon {
      color: #fcf82d;
    }

    &--skull {
      color: #ffffff;
    }

    &--fish {
      color: #2dc9fc;
    }
  }

  &--clickable {
    cursor: pointer;
  }
}
