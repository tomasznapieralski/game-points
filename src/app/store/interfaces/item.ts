export interface ItemInterface {
  icon: string;
  basePoints: number;
  bonusPoints: number;
  bonusMultiplier: number;
}

export interface CollectedItemInterface {
  item: ItemInterface;
  quantity: number;
}
