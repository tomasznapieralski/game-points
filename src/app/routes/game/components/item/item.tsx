import React, { useEffect } from 'react';
import { Action } from 'redux';
import classNames from 'classnames';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faBiohazard,
  faFire,
  faSpider,
  faLemon,
  faFish,
  faSkullCrossbones,
  IconDefinition,
} from '@fortawesome/free-solid-svg-icons';

import { ITEMS, ITEM_LIFE_TIME } from '../../../../constraints/items';

import './item.scss';

interface PropsInterface {
  icon: string;
  id?: number;
  clickHandler?: () => Action;
  initHandler?: (id: number) => Action;
  timeoutHandler?: (id: number) => Action;
}

const Item: React.FC<PropsInterface> = ({
  icon,
  id,
  clickHandler,
  initHandler,
  timeoutHandler,
}) => {
  let selectedIcon: IconDefinition;
  let iconClass: string;

  switch (icon) {
    case ITEMS.BIOHAZARD: {
      selectedIcon = faBiohazard;
      iconClass = 'item__icon--biohazard';
      break;
    }
    case ITEMS.FIRE: {
      selectedIcon = faFire;
      iconClass = 'item__icon--fire';
      break;
    }
    case ITEMS.SPIDER: {
      selectedIcon = faSpider;
      iconClass = 'item__icon--spider';
      break;
    }
    case ITEMS.LEMON: {
      selectedIcon = faLemon;
      iconClass = 'item__icon--lemon';
      break;
    }
    case ITEMS.SKULL: {
      selectedIcon = faSkullCrossbones;
      iconClass = 'item__icon--skull';
      break;
    }
    case ITEMS.FISH:
    default: {
      selectedIcon = faFish;
      iconClass = 'item__icon--fish';
      break;
    }
  }

  useEffect(() => {
    if (initHandler && id) {
      initHandler(id);
    }
  }, [initHandler, id]);

  useEffect(() => {
    setTimeout(() => {
      if (timeoutHandler && id) {
        timeoutHandler(id);
      }
    }, ITEM_LIFE_TIME);
  }, [timeoutHandler, id]);

  return (
    <div
      className={classNames('item', { 'item--clickable': clickHandler })}
      onClick={() => clickHandler && clickHandler()}
    >
      <FontAwesomeIcon
        className={classNames('item__icon', iconClass)}
        icon={selectedIcon}
      />
    </div>
  );
}

export default Item;
