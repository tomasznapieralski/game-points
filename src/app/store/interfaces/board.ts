import { ItemInterface } from './item';

export interface BoardSlotInterface {
  id: number;
  row: number;
  col: number;
  isHidden: boolean;
  item: ItemInterface | null;
}
