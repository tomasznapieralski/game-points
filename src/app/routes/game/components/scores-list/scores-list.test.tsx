import React from 'react';
import renderer from 'react-test-renderer';
import { render, cleanup } from '@testing-library/react';

import { CollectedItemInterface } from '../../../../store/interfaces/item';

import ScoresList from './scores-list';

afterEach(cleanup);

test('Snapshot Correct Component', () => {
  const tree = renderer
    .create(
      <ScoresList
        collectedItems={[]}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

test('Scores are counted correctly', () => {
  const collectedItems: CollectedItemInterface[] = [
    {
      item: {
        icon: 'fish',
        basePoints: 150,
        bonusPoints: 100,
        bonusMultiplier: 3,
      },
      quantity: 10,
    },
    {
      item: {
        icon: 'lemon',
        basePoints: 100,
        bonusPoints: 0,
        bonusMultiplier: 0,
      },
      quantity: 10,
    },
  ];

  const tree = renderer
    .create(
      <ScoresList
        collectedItems={collectedItems}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();

  const { container } = render(
    <ScoresList
      collectedItems={collectedItems}
    />
  );

  const items = container.getElementsByClassName('scores-list__item');
  expect(items[0].children[1].textContent).toBe('2500');
  expect(items[1].children[1].textContent).toBe('300');
  expect(items[2].children[1].textContent).toBe('2800');
});
