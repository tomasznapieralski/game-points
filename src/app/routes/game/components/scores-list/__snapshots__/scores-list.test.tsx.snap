// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scores are counted correctly 1`] = `
<ul
  className="scores-list"
>
  <li
    className="scores-list__item"
  >
    <div>
      Base points:
    </div>
    <div>
      2500
    </div>
  </li>
  <li
    className="scores-list__item"
  >
    <div>
      Bonus points:
    </div>
    <div>
      300
    </div>
  </li>
  <li
    className="scores-list__item"
  >
    <div>
      Total points:
    </div>
    <div>
      2800
    </div>
  </li>
</ul>
`;

exports[`Snapshot Correct Component 1`] = `
<ul
  className="scores-list"
>
  <li
    className="scores-list__item"
  >
    <div>
      Base points:
    </div>
    <div>
      0
    </div>
  </li>
  <li
    className="scores-list__item"
  >
    <div>
      Bonus points:
    </div>
    <div>
      0
    </div>
  </li>
  <li
    className="scores-list__item"
  >
    <div>
      Total points:
    </div>
    <div>
      0
    </div>
  </li>
</ul>
`;
